x-pack/plugins/case/server/client/comments/add.ts

```diff
@@ -10,135 +10,297 @@ import { pipe } from 'fp-ts/lib/pipeable';
+  buildCaseUserActionItem,
+  buildCommentUserActionItem,
+} from '../../services/user_actions/helpers';
+
+import { CaseServiceSetup, CaseUserActionServiceSetup } from '../../services';
+import { CommentableCase } from '../../common';
+import { CaseClientImpl } from '..';
+
+async function getSubCase({
+  caseService,
+  savedObjectsClient,
+  caseId,
+  createdAt,
+  userActionService,
+  user,
+}: {
+  caseService: CaseServiceSetup;
+  savedObjectsClient: SavedObjectsClientContract;
+  caseId: string;
+  createdAt: string;
+  userActionService: CaseUserActionServiceSetup;
+  user: User;
+}): Promise<SavedObject<SubCaseAttributes>> {
+  const mostRecentSubCase = await caseService.getMostRecentSubCase(savedObjectsClient, caseId);
+  if (mostRecentSubCase && mostRecentSubCase.attributes.status !== CaseStatuses.closed) {
+    return mostRecentSubCase;
+  }
 
-import { CaseClientAddComment, CaseClientFactoryArguments } from '../types';
-import { CASE_SAVED_OBJECT } from '../../saved_object_types';
+  const newSubCase = await caseService.createSubCase({
+    client: savedObjectsClient,
+    createdAt,
+    caseId,
+    createdBy: user,
+  });
+  await userActionService.postUserActions({
+    client: savedObjectsClient,
+    actions: [
+      buildCaseUserActionItem({
+        action: 'create',
+        actionAt: createdAt,
+        actionBy: user,
+        caseId,
+        subCaseId: newSubCase.id,
+        fields: ['status', 'sub_case'],
+        newValue: JSON.stringify({ status: newSubCase.attributes.status }),
+      }),
+    ],
+  });
+  return newSubCase;
+}
+
+interface AddCommentFromRuleArgs {
+  caseClient: CaseClientImpl;
+  caseId: string;
+  comment: CommentRequestAlertType;
+  savedObjectsClient: SavedObjectsClientContract;
+  caseService: CaseServiceSetup;
+  userActionService: CaseUserActionServiceSetup;
+}
 
-export const addComment = ({
+const addGeneratedAlerts = async ({
   savedObjectsClient,
   caseService,
   userActionService,
-  request,
-}: CaseClientFactoryArguments) => async ({
   caseClient,
   caseId,
   comment,
-}: CaseClientAddComment): Promise<CaseResponse> => {
+}: AddCommentFromRuleArgs): Promise<CollectionWithSubCaseResponse> => {
   const query = pipe(
-    CommentRequestRt.decode(comment),
+    AlertCommentRequestRt.decode(comment),
     fold(throwErrors(Boom.badRequest), identity)
   );
 
-  decodeComment(comment);
+  decodeCommentRequest(comment);
+
+  // This function only supports adding generated alerts
+  if (comment.type !== CommentType.generatedAlert) {
+    throw Boom.internal('Attempting to add a non generated alert in the wrong context');
+  }
+  const createdDate = new Date().toISOString();
 
-  const myCase = await caseService.getCase({
+  const caseInfo = await caseService.getCase({
     client: savedObjectsClient,
-    caseId,
+    id: caseId,
   });
 
-  // An alert cannot be attach to a closed case.
-  if (query.type === CommentType.alert && myCase.attributes.status === CaseStatuses.closed) {
```

Comment:
Do we want to keep this check though?

Reply:
Yeah, it should be in `CommentableCase` now haha.